sunpy/map/tests/test_mapbase.py

```diff
@@ -821,15 +821,78 @@ def test_quicklook(aia171_test_map):
         assert aia171_test_map._repr_html_() in html_string
 
 
-def test_submap_non_kwarg_parameters(aia171_test_map):
-    bl = SkyCoord(-300*u.arcsec, -300*u.arcsec, frame=aia171_test_map.coordinate_frame)
-    tr = SkyCoord(500*u.arcsec, 500*u.arcsec, frame=aia171_test_map.coordinate_frame)
+def test_submap_inputs(generic_map):
+    bl_pix = [3, 2] * u.pix
+    bl_coord = SkyCoord(20, -10, unit=u.arcsec,
+                        frame=generic_map.coordinate_frame)
+
+    tr_pix = [5, 4] * u.pix
+    tr_coord = SkyCoord(0, 10, unit=u.arcsec,
+                        frame=generic_map.coordinate_frame)
+
+    bl_tr_coord = SkyCoord([20, 0], [-10, 10], unit=u.arcsec,
+                           frame=generic_map.coordinate_frame)
+
+    width_pix = 2 * u.pix
+    height_pix = 2 * u.pix
+    width_deg = 20 * u.arcsec
+    height_deg = 20 * u.arcsec
+
+    # works
+    smap = generic_map.submap(bl_coord, top_right=tr_coord)
+    # assert u.allclose(smap.dimensions, (1, 1) * u.pix)
+    smap = generic_map.submap(bl_coord, width=width_deg, height=height_deg)
+    # assert u.allclose(smap.dimensions, (1, 1) * u.pix)
+    smap = generic_map.submap(bl_tr_coord)
+    # assert u.allclose(smap.dimensions, (1, 1) * u.pix)
+
+    # works
+    smap = generic_map.submap(bl_pix, top_right=tr_pix)
+    # assert u.allclose(smap.dimensions, (1, 1) * u.pix)
+    # smap = generic_map.submap(bl_pix, width=width_pix, height=height_pix)
+    # assert u.allclose(smap.dimensions, (1, 1) * u.pix)
+
+    # deprecated
+    with pytest.warns(FutureWarning):
+        smap = generic_map.submap(bl_coord, tr_coord)
+    # assert u.allclose(smap.dimensions, (1, 1) * u.pix)
 
     with pytest.warns(FutureWarning):
-        aia171_test_map.submap(bl, tr)
+        smap = generic_map.submap(bl_pix, tr_pix)
+    # assert u.allclose(smap.dimensions, (1, 1) * u.pix)
 
-    width = 10 * u.arcsec
-    height = 10 * u.arcsec
+    # error
+    with pytest.raises(TypeError):
+        generic_map.submap(bl_coord, width_deg, height_deg)
+    with pytest.raises(TypeError):
+        generic_map.submap(bl_pix, width_pix, height_pix)
 
-    with pytest.warns(FutureWarning):
-        aia171_test_map.submap(bl, width, height)
+    # TODO: The error message raised here is poor
+    with pytest.raises(TypeError):
+        generic_map.submap(bl_coord, width_deg, height=height_deg)
+
+    # TODO: The error raised here is completely wrong!
+    with pytest.raises(TypeError):
+        generic_map.submap(bl_pix, width_pix, height=height_pix)
+
```

Comment:
We were facing a similar problem in draw_rect
I fixed it with the following code:
```
        if isinstance(top_right, u.Quantity) and isinstance(width, u.Quantity):
            # The decorator assigns the first positional arg to top_right and so on.
            height = width
            width = top_right
            top_right = None
```
As far as the error message, we would need to modify the decorator. I don't see resolving this without losing some generality. We could make a specific decorator that has specific type checks related to input types here.
Then calling it above this one to fix the inputs before we check the positional args.

Reply:
This *should* error, it's not supported, it's just the errors it raises are not helpful lol. Where as for `draw_rectangle` we should support it.

Reply:
This is going to be painful.
How do you suppose we fix the error message?

Reply:
I don't see this happening without making a decorator specific to submap.
We to check if there is only one positional arg, then it should be either be a pix array of size two or a skycoord.
If two positional, then both should be u.arcsec or u.pix.
If all thee, we thow an error.